src/finn/custom_op/fpgadataflow/iodma.py

```diff
@@ -0,0 +1,189 @@
+            "direction": ("s", False, "in"),
+            # shape describing input vecs per execution
+            "numInputVectors": ("ints", False, [1]),
+        }
+        my_attrs.update(super().get_nodeattr_types())
+        return my_attrs
+
+    def get_normal_input_shape(self):
+        vecs = list(self.get_nodeattr("numInputVectors"))
+        num_ch = self.get_nodeattr("NumChannels")
+        ishape = tuple(vecs + [num_ch])
+        return ishape
+
+    def get_normal_output_shape(self):
+        return self.get_normal_input_shape()
+
+    def get_folded_input_shape(self):
+        shape = list(self.get_normal_input_shape())
+        itype_bits = self.get_input_datatype().bitwidth()
+        intfw = self.get_nodeattr("intfWidth")
+        elems_per_word = intfw / itype_bits
+        fold_depth = round(shape[-1] / elems_per_word)
+        shape[-1] = fold_depth
+        shape.append(elems_per_word)
+        return tuple(shape)
+
+    def get_folded_output_shape(self):
+        return self.get_folded_input_shape()
+
+    def make_shape_compatible_op(self, model):
+        exp_ishape = self.get_normal_input_shape()
+        oshape = self.get_normal_output_shape()
+        ishape = tuple(model.get_tensor_shape(self.onnx_node.input[0]))
+        assert ishape == exp_ishape, "Unexpected input shape."
+        # implement tensor with correct shape
+        values = np.random.randn(*oshape).astype(np.float32)
+        return helper.make_node(
+            "Constant",
+            inputs=[],
+            outputs=[self.onnx_node.output[0]],
+            value=helper.make_tensor(
+                name="const_tensor",
+                data_type=TensorProto.FLOAT,
+                dims=values.shape,
+                vals=values.flatten().astype(float),
+            ),
+        )
+
+    def infer_node_datatype(self, model):
+        node = self.onnx_node
+        # data type stays the same
+        dtype = model.get_tensor_datatype(node.input[0])
+        exp_idtype = self.get_input_datatype()
+        assert dtype == exp_idtype, "Unexpected datatype."
+        model.set_tensor_datatype(node.output[0], dtype)
+
+    def verify_node(self):
+        pass
+
+    def get_input_datatype(self):
+        """Returns FINN DataType of input."""
+        return DataType[self.get_nodeattr("dataType")]
+
+    def get_output_datatype(self):
+        """Returns FINN DataType of output. (Same as input datatype)"""
+        return self.get_input_datatype()
+
+    def get_instream_width(self):
+        return self.get_nodeattr("intfWidth")
+
+    def get_outstream_width(self):
+        return self.get_instream_width()
+
+    def get_number_output_values(self):
+        oshape = self.get_normal_output_shape()
+        itype_bits = self.get_input_datatype().bitwidth()
+        intfw = self.get_nodeattr("intfWidth")
+        nelems = np.prod(oshape)
+        nbits = nelems * itype_bits
+        assert nbits % intfw == 0, "DMA: total transfer size must be word multiple"
```

Comment:
so from this line it looks like there is an alignment assumption/requirement s.t. the total transfer size is divisible by the interface width (btw np.prod on the packed output shape may be quicker for the num elems calculation here). 

1) how do we deal with cases where the first/last layers of a network may produce a nonaligned number of bits?
2) are there any extra considerations needed for AXI stream width? (internal stream widths always get padded to a multiple of 8, normally not a problem since the padded bits are just ignored in the next layer)

Reply:
Documented transfer size requirements with relation to intfWidth/streamWidth. This version of the hlslib dma (and therefore the IODMA node itself) won't work with streams widths which aren't a multiple of 8b and tensors whose total number of bits is not divisible by the AXI-MM interface width. This shouldn't be a problem for network inputs or outputs, but may be problematic for streaming weights later on. I would propose modifying the hlslib dma and the IODMA when use-cases require.